lightning/src/routing/router.rs

```diff
@@ -0,0 +1,1186 @@
+								if chan.one_to_two.enabled {
+									add_entry!(chan_id, chan.two_to_one.src_node_id, chan.one_to_two, chan.features, $fee_to_target_msat);
+								}
+							}
+						}
+					}
+				}
+			}
+		};
+	}
+
+	match network.nodes.get(target) {
+		None => {},
+		Some(node) => {
+			add_entries_to_cheapest_to_target_node!(node, target, 0);
+		},
+	}
+
+	for hop in last_hops.iter() {
+		if first_hops.is_none() || hop.src_node_id != network.our_node_id { // first_hop overrules last_hops
+			if network.nodes.get(&hop.src_node_id).is_some() {
+				if first_hops.is_some() {
+					if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+						// Currently there are no channel-context features defined, so we are a
+						// bit lazy here. In the future, we should pull them out via our
+						// ChannelManager, but there's no reason to waste the space until we
+						// need them.
+						add_entry!(first_hop, hop.src_node_id, dummy_directional_info, ChannelFeatures::with_known_relevant_init_flags(&features), 0);
+					}
+				}
+				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+				// really sucks, cause we're gonna need that eventually.
+				add_entry!(hop.short_channel_id, target, hop, ChannelFeatures::empty(), 0);
+			}
+		}
+	}
+
+	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
+		if pubkey == network.our_node_id {
+			let mut res = vec!(dist.remove(&network.our_node_id).unwrap().3);
+			loop {
+				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
+					res.last_mut().unwrap().node_features = NodeFeatures::with_known_relevant_init_flags(&features);
+				} else if let Some(node) = network.nodes.get(&res.last().unwrap().pubkey) {
+					res.last_mut().unwrap().node_features = node.features.clone();
+				} else {
+					// We should be able to fill in features for everything except the last
+					// hop, if the last hop was provided via a BOLT 11 invoice (though we
+					// should be able to extend it further as BOLT 11 does have feature
+					// flags for the last hop node itself).
+					assert!(res.last().unwrap().pubkey == *target);
+				}
+				if res.last().unwrap().pubkey == *target {
+					break;
+				}
+
+				let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
+					Some(hop) => hop.3,
+					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination", action: ErrorAction::IgnoreError}),
+				};
+				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
+				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
+				res.push(new_entry);
+			}
+			res.last_mut().unwrap().fee_msat = final_value_msat;
+			res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
+			let route = Route { paths: vec![res] };
+			// log_trace!(self, "Got route: {}", log_route!(route));
```

Comment:
I have no idea how this thing works and why it fails right now :)

Reply:
our logging stuff generally assumes its being called from a method that takes an &self (ie a non-static method) and then accesses self.logger. While we should probably update our macros to make that not be the case, for now you can use the hack we have at https://github.com/rust-bitcoin/rust-lightning/blob/master/lightning/src/ln/onion_utils.rs#L443.